guides/hack/11-built-in-types/34-enum.md

```diff
@@ -0,0 +1,197 @@
+Use an enum (enumerated type) to create a set of named, constant, immutable values. 
+
+In Hack, enums are limited to `int` or `string` (as an [`Arraykey`](/hack/built-in-types/arraykey)), or other `enum` values.
+
+## Quickstart
+To access an enum's value, use its full name, as in `Colors::Blue` or `Permission::Read`.
+
+```Colors.hack no-auto-output
+enum Colors: int {
+  Red = 3;
+  Green = 5;
+  Blue = 10;
+  Default = 3; // duplicate value is okay
+}
+```
+
+```Permission.hack no-auto-output
+enum Permission: string {
+  Read = 'R';
+  Write = 'W';
+  Execute = 'E';
+  Delete = 'D';
+}
+```
+
+Additionally, by using the [`as`](/hack/expressions-and-operators/type-assertions#enforcing-types-with-as-and-as) operator to enforce type, you can initialize your enum with static expressions that reference other enum values.
+
+```BitFlags.hack no-auto-output
+enum BitFlags: int as int {
+  F1 = 1; // value 1
+  F2 = BitFlags::F1 << 1; // value 2
+  F3 = BitFlags::F2 << 1; // value 4
+  F4 = 4 + 3; // value 7
+}
+```
+
+## Full Example
+With an enum, we can create a placement-direction system with names like `Top`, `Bottom`, `Left`, `Right`, and `Center`, then direct output accordingly to write text to the top, bottom, left, right, or center of a window.
+
+```Positions.inc.hack no-auto-output
+enum Position: int {
+  Top = 0;
+  Bottom = 1;
+  Left = 2;
+  Right = 3;
+  Center = 4;
+}
+
+function writeText(string $text, Position $pos): void {
+  switch ($pos) {
+    case Position::Top:
+      // ...
+      break;
+    case Position::Center:
+      // ...
+      break;
+    case Position::Right:
+      // ...
+      break;
+    case Position::Left:
+      // ...
+      break;
+    case Position::Bottom:
+      // ...
+      break;
+  }
+}
+
+<<__EntryPoint>>
+function main(): void {
+  writeText("Hello", Position::Bottom);
+  writeText("Today", Position::Left);
+}
+```
+
+## Library Methods
+All enums implement these public static methods.
+
+### `getNames() / getValues()`
+Returns a map-like array of enum constant values and their names.
+
+```NamesValues.hack
+enum Position: int {
+  Top = 0;
+  Bottom = 1;
+  Left = 2;
+  Right = 3;
+  Center = 4;
+}
+
+<<__EntryPoint>>
+function main(): void {
+  $names = Position::getNames();
+  echo " Position::getNames() ---\n";
+  foreach ($names as $key => $value) {
+    echo "\tkey >$key< has value >$value<\n";
+  }
+
+  $values = Position::getValues();
+  echo "Position::getValues() ---\n";
+  foreach ($values as $key => $value) {
+    echo "\tkey >$key< has value >$value<\n";
+  }
+}
+```
+
+### `assert()` / `coerce()`
+`assert($value)` checks if a value exists in an enum, and if it does, returns the value; if the value does not exist, throws an `UnexpectedValueException`.
+
+`coerce($value)` checks if a value exists in an enum, and if it does, returns the value; if the value does not exist, returns `null`.
```

Comment:
Both assert and coerce may cast the value given from int to string if no direct int match is found. The same goes for strings that are canonical representations of ints. They will be cast to an integer if no string match is found. This is maybe expected for coerce, but noteworthy and strange for assert.

Reply:
I can add as:

![image](https://user-images.githubusercontent.com/5179225/145281174-4d050570-b293-420a-a2a5-13511ec1bf99.png)

Reply:
My description was incomplete and this incompleteness leaked into your description as well. The cast string -> int / int -> string value is only returned if the result of the cast is a member of the enum. The current description hint towards unconditionally returning the result of the cast. A more accurate description would be:

Both library methods will, when no exact match is found, attempt to do a cast to the other arraykey type. If the cast is not reversible / lossless, or the resulting value is still not a member of the enum after the cast, the failure result occurs. Where failure for assert is throwing a UnexpectedValueException and for coerce it is returning null.

Reply:
And string in the format of a number is a wider description than and intish string. "3.5" is a string in the format of a number, but not an intish string.

Reply:
Yeah, I agree that it is too wide of a description. Was writing from a flawed viewpoint earlier, but has been updated.

Going to use your second paragraph above as a note with some slight wording changes. Thanks!